### PR TITLE
refactor external client to make single getAWSConfig call per connect 

### PR DESCRIPTION
**Pull request by erhancagirici**
### Description of your changes

Refactors `getAWSConfig` calls in the external client code, to make single call per connect. `getAWSconfig` calls involve resolving credentials and executing AWS STS calls when there are relevant `AssumeRole` or `AssumeRoleWithWebIdentity`  operations needed. This was being called multiple times per connect, causing superfluous invocations.

This PR will improve the situation reported at #997

I have:

- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
Via uptest: https://github.com/upbound/provider-aws/actions/runs/7141372941

### CODE REVIEW

**Comment by haarchri** [comment]
Is it possible to implement credential caching? Currently, it seems we make a single call per connect. However, when managing multiple resources, each resource triggers a separate call. Considering that IRSA credentials are valid for 15 minutes, it would be more efficient to perform the call only once every 15 minutes, even when dealing with numerous managed resources - what do you think ?

**Comment by erhancagirici** [comment]
@haarchri This PR is an initial improvement for addressing the low-hanging fruit. Our plan is to make further improvements in iterations. This is especially motivated by the fact that those parts of the provider are currently fractured because we have a hybrid architecture for that provider (both the CLI-based & SDK-based upjet architectures are in use at the moment). 
The plan is once we completely switch to the nofork arch for the remaining few resources , we will be able to simplify the provider setup logic there. And as we simplify provider setup logic, it will be easier for us to make further optimizations like credential caching, that is definitely in our scope.

**Comment by erhancagirici** [comment]
/test-examples="examples/iam/role.yaml"

File: internal/clients/aws.go [hunk 1]
```diff
@@ -199,7 +196,7 @@ func DefaultTerraformSetupBuilder(ctx context.Context, c client.Client, mg resou
 	if pc.Spec.Endpoint != nil {
 		if pc.Spec.Endpoint.URL.Static != nil {
 			if len(pc.Spec.Endpoint.Services) > 0 && *pc.Spec.Endpoint.URL.Static == "" {
-				return errors.Wrap(err, "endpoint is wrong")
```
Comment (ulucinar) [inline internal/clients/aws.go:202]: Good catch! Then, we were previously silently non-erroring :)

**Review by ulucinar — APPROVED** [review]
Thanks @erhancagirici, lgtm.

**Pull request merged**